export type AnimInfo = {
    Step: number;
    Group: number;
};
export type PropertiesType = {
    TextXAlignment: Enum.TextXAlignment?;
    TextYAlignment: Enum.TextYAlignment?;
    TextColor3: Color3?;
    Font: Enum.Font?;
    TextScaled: boolean?;
    Scale: number?;
    TextStrokeColor3: Color3?;
    StrokeTransparency: number?;
    TextTransparency: number?;
    Overfill: boolean?;
    Step: number?;
    Group: number?;
    TextSize: number?;
    TextWrapped: boolean?;
};
return {
 AnimInfo = nil;
 PropertiesType = nil;   
};